### PR TITLE
Add strong root logging and $vm helpers for memory leak debugging

### PR DESCRIPTION
**Pull request by justinmichaud**
#### 41dc316f40f48a3a1394f1ef811839390d06e75c
<pre>
Add strong root logging and $vm helpers for memory leak debugging
<a href="https://bugs.webkit.org/show_bug.cgi?id=277835">https://bugs.webkit.org/show_bug.cgi?id=277835</a>

Reviewed by David Degazio.

This patch adds some new options for debugging memory leaks. It also
allows $vm to be enabled again on GTK developer builds.

For example, we can debug a memory leak in the JSC heap by:

1) Run:
```
JSC_validateOptions=1 JSC_useDollarVM=1 JSC_enableStrongRefTracker=1 JSC_dumpHeapOnLowMemory=1 JSC_useSourceProviderCache=0 JSC_useCodeCache=0 WEBKIT_DISABLE_SANDBOX_THIS_IS_DANGEROUS=1 Tools/Scripts/run-minibrowser --gtk --debug <a href="https://webkit.org">https://webkit.org</a>
```

Alternatively, to track C allocations, disable system malloc in CMakeLists.txt (I recommend adding a #if !USE_SYSTEM_MALLOC #error #endif to confirmn), run:

```
JSC_useSourceProviderCache=0 JSC_useCodeCache=0 WEBKIT_DISABLE_SANDBOX_THIS_IS_DANGEROUS=1 MALLOC=0 WEB_PROCESS_CMD_PREFIX=&quot;/usr/bin/valgrind --tool=massif &quot; Tools/Scripts/run-minibrowser --gtk --debug <a href="https://webkit.org">https://webkit.org</a>
```

Then, to release memory, dump the JSC heap, and log all remaining Strong&lt;&gt; references, run:

```
javascript:$vm.gc()
javascript:$vm.triggerMemoryPressure()
```

You should get a gc snapshot file. To graph it (or just do dump it to a human-readable format), run:
```
Tools/Scripts/heap-snapshot-to-graphviz.js /tmp/GCHeapM8FVBp  ~/
dot -Tsvg ~/GCDebugging.gv.txt -O
```

You can edit `heap-snapshot-to-graphviz.js` to print root paths, and there are many other graph-related things you can do in js.

You can also open this file in WebInspector by adding a JavaScript allocations track, and clicking the second import button (not the main import button).

If you see that your heap objects are kept alive by StrongReferences, then the strong logging may point to which C++ object is keeping them alive, and you can continue your investigation from there.

As a general tip, you should ensure instrumentingAgents.inspectorEnvironment().developerExtrasEnabled() is false on your platform when diagnosing these bugs, since WebInspector can keep things alive unnecesarily.

It may also be helpful to disable Process Swap on Naviagation.

* Source/JavaScriptCore/heap/Heap.cpp:
(JSC::Heap::addCoreConstraints):
(JSC::StrongLoggingDisabledScope::StrongLoggingDisabledScope):
(JSC::StrongLoggingDisabledScope::~StrongLoggingDisabledScope):
(JSC::debugReportLiveStrong):
(JSC::debugReportDeadStrong):
(JSC::debugLogAllLiveStrongs):
* Source/JavaScriptCore/heap/Heap.h:
* Source/JavaScriptCore/heap/MarkedSpace.h:
* Source/JavaScriptCore/heap/MarkedSpaceInlines.h:
(JSC::MarkedSpace::forEachWeakInParallel):
* Source/JavaScriptCore/heap/Strong.h:
(JSC::Strong::Strong):
(JSC::Strong::~Strong):
* Source/JavaScriptCore/heap/StrongInlines.h:
(JSC::shouldStrongDestructorGrabLock&gt;::Strong):
(JSC::shouldStrongDestructorGrabLock&gt;::set):
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::VM::VM):
* Source/JavaScriptCore/tools/JSDollarVM.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::JSDollarVM::finishCreation):
* Source/WTF/wtf/MemoryPressureHandler.cpp:
(WTF::MemoryPressureHandler::setShouldUsePeriodicMemoryMonitor):
* Source/WebCore/bindings/js/GCController.h:
* Source/WebInspectorUI/UserInterface/Workers/HeapSnapshot/HeapSnapshot.js:
(HeapSnapshot):
(HeapSnapshot.prototype.shortestGCRootPath):
(HeapSnapshot.prototype.serializeEdge):
(HeapSnapshot.prototype._determineGCRootPaths):
* Source/WebKit/NetworkProcess/soup/NetworkProcessSoup.cpp:
(WebKit::NetworkProcess::platformInitializeNetworkProcess):
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::initializeWebProcess):
* Source/cmake/OptionsGTK.cmake:
* Tools/Scripts/heap-snapshot-to-graphviz.js: Added.
(true.console.log.string_appeared_here.escapeOutput):
(return.unsafe.replace):

Canonical link: <a href="https://commits.webkit.org/284791@main">https://commits.webkit.org/284791@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/55aa5bf44292a913eaea3eec38691d38b90493cf

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/70489 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/49895 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/23254 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/74578 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/21667 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/72606 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/57695 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/21507 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/55864 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/14325 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/73555 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/45381 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/60766 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/36326 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/42040 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/18202 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/20028 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/63610 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/63972 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/18546 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/76297 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/69736 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/14715 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/17784 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/63586 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/14758 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/60833 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/63520 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15615 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/11557 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/5192 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/91519 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/45697 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/464 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/19952 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/46771 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/48048 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/46513 "Built successfully") | | |

### CODE REVIEW

File: Source/WebKit/WebProcess/WebProcess.cpp [hunk 1]
```diff
@@ -468,7 +471,9 @@ void WebProcess::initializeWebProcess(WebProcessCreationParameters&& parameters,
 #if ENABLE(PERIODIC_MEMORY_MONITOR)
         if (auto pollInterval = parameters.memoryFootprintPollIntervalForTesting)
             memoryPressureHandler.setMemoryFootprintPollIntervalForTesting(pollInterval);
-        memoryPressureHandler.setShouldUsePeriodicMemoryMonitor(true);
+        // If we're running with FastMalloc disabled, some kind of testing or debugging is probably happening.
+        // Let's be nice and not enable the memory kill mechanism.
```
Comment (mcatanzaro) [inline Source/WebKit/WebProcess/WebProcess.cpp:478]: This is probably only true if `USE(SYSTEM_MALLOC)` is false. If system malloc is enabled at build time then you're just detecting that bmalloc isn't supported.

File: Source/cmake/OptionsGTK.cmake [hunk 2]
```diff
@@ -51,6 +51,7 @@ WEBKIT_OPTION_DEFINE(USE_LIBDRM "Whether to enable usage of libdrm." PUBLIC ON)
 WEBKIT_OPTION_DEFINE(USE_LIBHYPHEN "Whether to enable the default automatic hyphenation implementation." PUBLIC ON)
 WEBKIT_OPTION_DEFINE(USE_LIBSECRET "Whether to enable the persistent credential storage using libsecret." PUBLIC ON)
 WEBKIT_OPTION_DEFINE(USE_SOUP2 "Whether to enable usage of Soup 2 instead of Soup 3." PUBLIC OFF)
+WEBKIT_OPTION_DEFINE(ENABLE_JSC_RESTRICTED_OPTIONS_BY_DEFAULT "Whether to enable dangerous development options in JSC by default." PUBLIC OFF)
```
Comment (mcatanzaro) [inline Source/cmake/OptionsGTK.cmake]: Looks like this is unused leftover and can be removed?

(But if it was not unused, I would next ask why it should be public if it's a dangerous development option, and also why it needs to be specific to GTK port instead of going into WebKitFeatures.cmake.)

Comment (justinmichaud) [reply Source/cmake/OptionsGTK.cmake]: The Apple ports enable this for debug builds automatically now, so it really is specific to our ports. Really, it should be on by default for development builds, and off everywhere else. What is the best way to accomplish that?

File: Source/WebKit/WebProcess/WebProcess.cpp (hunk 1 above)
Comment (justinmichaud) [reply Source/WebKit/WebProcess/WebProcess.cpp:478]: I copied and pasted this from the file below, but actually that makes a lot of sense, and might explain memory issues with MotionMark on ArmV7

File: Source/cmake/OptionsGTK.cmake (hunk 2 above)
Comment (kmiller68) [reply Source/cmake/OptionsGTK.cmake]: Technically, it's on for all development builds (i.e. Release/Debug but not Production/Profiling) on Apple ports. Not sure if that's particularly important but I thought I'd mention it.

Comment (mcatanzaro) [reply Source/cmake/OptionsGTK.cmake]: > The Apple ports enable this for debug builds automatically now

No they don't. A `git grep` indicates `ENABLE_JSC_RESTRICTED_OPTIONS_BY_DEFAULT` does not exist. :)

> Really, it should be on by default for development builds, and off everywhere else. What is the best way to accomplish that?

In that case, doing it in OptionsGTK.cmake and OptionsWPE.cmake really is required, because `DEVELOPER_MODE` is a platform-specific option. But please don't make it `PUBLIC`.

File: Source/JavaScriptCore/tools/JSDollarVM.cpp [hunk 3]
```diff
@@ -63,6 +63,7 @@
 #include "VMTrapsInlines.h"
 #include "WasmCapabilities.h"
 #include <unicode/uversion.h>
+#include <wtf/MemoryPressureHandler.h>
```
Comment (mcatanzaro) [inline Source/JavaScriptCore/tools/JSDollarVM.cpp]: Please alphabetize this.

**Comment by MenloDorian** [comment]
> This patch adds some new options for debugging memory leaks

Can you define what kind of "memory leaks" you are referring to here?
(1) C/C++ code memory leaks?
(2) JS heap "memory leaks" due to JS code retaining memory longer than needed?

File: Source/WebInspectorUI/UserInterface/Workers/HeapSnapshot/HeapSnapshot.js [hunk 4]
```diff
@@ -761,11 +773,26 @@ HeapSnapshot = class HeapSnapshot
             pathsBeingProcessed[i] = undefined;
 
             if (this._nodeOrdinalIsGCRoot[nodeOrdinal]) {
-                let fullPath = currentPath.slice();
                 let nodeIndex = nodeOrdinal * this._nodeFieldCount;
-                fullPath.push({node: nodeIndex});
-                gcRootPaths.push(fullPath);
-                continue;
+
```
Comment (justinmichaud) [inline Source/WebInspectorUI/UserInterface/Workers/HeapSnapshot/HeapSnapshot.js:779]: @dcrousso How can I test that this doesn't break WebInspector?

File: Source/JavaScriptCore/heap/Heap.cpp [hunk 5]
```diff
@@ -3328,4 +3328,61 @@ DEFINE_DYNAMIC_ISO_SUBSPACE_MEMBER_SLOW(moduleProgramExecutableSpace)
...
+
+uintptr_t debugReportLiveStrong()
+{
+    if (!Options::logLiveStrongs() || strongLoggingDisabled.load() > 0)
+        return 0;
+    Locker locker(strongMapLock);
+    if (!debugStrongMap)
+        debugStrongMap = new HashMap<uintptr_t, String>();
+    WTF::StringPrintStream os;
+    WTFReportBacktraceWithPrefixAndPrintStream(os, "");
```
Comment (rreno) [inline Source/JavaScriptCore/heap/Heap.cpp]: Can you use `WTF::StackShot` instead of a string?

Then the map is `HashMap<token, std::unique_ptr<StackShot>>`

File: Source/JavaScriptCore/heap/Strong.h [hunk 6]
```diff
@@ -144,6 +160,10 @@ template <typename T, ShouldStrongDestructorGrabLock shouldStrongDestructorGrabL
         HandleSet::heapFor(slot())->template writeBarrier<std::is_base_of_v<JSCell, T>>(slot(), value);
         *slot() = value;
     }
+
+#if ASSERT_ENABLED
+    uintptr_t strongID;
```
Comment (rreno) [inline Source/JavaScriptCore/heap/Strong.h]: Are `Strong` handles unique? Could you just use the address of `this` or the slot as the token instead?

**Review by rreno — COMMENTED** [review]
I've had a lot of success with a technique similar to this. Would be great to have it in a build instead of living on rotting branches!

File: Source/JavaScriptCore/heap/Heap.cpp [hunk 7]
```diff
@@ -3328,4 +3328,61 @@ DEFINE_DYNAMIC_ISO_SUBSPACE_MEMBER_SLOW(moduleProgramExecutableSpace)
...
+    --strongLoggingDisabled;
+}
+
+uintptr_t debugReportLiveStrong()
+{
+    if (!Options::logLiveStrongs() || strongLoggingDisabled.load() > 0)
+        return 0;
+    Locker locker(strongMapLock);
+    if (!debugStrongMap)
+        debugStrongMap = new HashMap<uintptr_t, String>();
```
Comment (rreno) [inline Source/JavaScriptCore/heap/Heap.cpp]: This could be useful in WebCore, especially since we want to get rid of all uses of Strong there. Could this hash map instead live in `WTF` so we can get access to it from WebCore too?

I use this object on my branch: https://github.com/rreno/WebKit/blob/eng/ref-tracking/Source/WTF/wtf/RefTracker.h

and I give access to the hash map globally via the static getters.

File: Source/JavaScriptCore/heap/Strong.h (hunk 6 above)
Comment (justinmichaud) [reply Source/JavaScriptCore/heap/Strong.h]: They get moved, unfortunately. I originally did this.

Comment (justinmichaud) [reply Source/JavaScriptCore/heap/Strong.h]: Maybe I could investigate why Vector wasn't running my destructor, I just did this as an easy solution.

File: Source/WebInspectorUI/UserInterface/Workers/HeapSnapshot/HeapSnapshot.js [hunk 8]
```diff
@@ -298,6 +307,8 @@ HeapSnapshot = class HeapSnapshot
         console.assert("node" in shortestPath[shortestPath.length - 1], "Path should end with a node");
 
         return shortestPath.map((component) => {
+            if (component.node && component.rootReason)
+                return Object.assign(this.serializeNode(component.node), { rootReason: component.rootReason });
```
Comment (dcrousso) [inline Source/WebInspectorUI/UserInterface/Workers/HeapSnapshot/HeapSnapshot.js]: Style: we dont add spaces after `{` or before `}` in Web Inspector

File: Source/WebInspectorUI/UserInterface/Workers/HeapSnapshot/HeapSnapshot.js [hunk 9]
```diff
@@ -298,6 +307,8 @@ HeapSnapshot = class HeapSnapshot
...
             if (component.node)
                 return this.serializeNode(component.node);
```
Comment (dcrousso) [inline Source/WebInspectorUI/UserInterface/Workers/HeapSnapshot/HeapSnapshot.js]: i think a cleaner way of doing this that doesnt involve `Object.assign` would be
```js
if (component.node) {
    let node = this.serializeNode(component.node);
    if (component.rootReason)
        node.rootReason = component.rootReason;
    return node;
}
```

File: Source/WebInspectorUI/UserInterface/Workers/HeapSnapshot/HeapSnapshot.js (hunk 4 above)
Comment (dcrousso) [reply Source/WebInspectorUI/UserInterface/Workers/HeapSnapshot/HeapSnapshot.js:779]: probably just open the Timelines Tab and take a heap snapshot and see if it loads and works correctly.  at the minimum id suggest checking that all the features mentioned in <https://webkit.org/web-inspector/timelines-tab/#javascript-allocations-timeline> and <https://webkit.org/blog/6425/memory-debugging-with-web-inspector/> (especially the part about determining/showing the shortest path) still work correctly

File: Source/WebInspectorUI/UserInterface/Workers/HeapSnapshot/HeapSnapshot.js [hunk 10]
```diff
@@ -761,11 +773,26 @@ HeapSnapshot = class HeapSnapshot
...
+                        continue
+                    rootReason = reason
+                    break
+                }
+
+                currentPath.push({node: nodeIndex, rootReason});
+
+                // These aren't really roots. They will only mark things that are already reachable.
+                if (rootReason != "WeakHandles") {
+                    gcRootPaths.push(currentPath);
```
Comment (dcrousso) [inline Source/WebInspectorUI/UserInterface/Workers/HeapSnapshot/HeapSnapshot.js]: we might still need the `currentPath.slice()` as otherwise we could be saving the value which may get modified in a subsequent iteration

File: Source/WebInspectorUI/UserInterface/Workers/HeapSnapshot/HeapSnapshot.js [hunk 11]
```diff
@@ -761,11 +773,26 @@ HeapSnapshot = class HeapSnapshot
             pathsBeingProcessed[i] = undefined;
 
             if (this._nodeOrdinalIsGCRoot[nodeOrdinal]) {
-                let fullPath = currentPath.slice();
                 let nodeIndex = nodeOrdinal * this._nodeFieldCount;
-                fullPath.push({node: nodeIndex});
-                gcRootPaths.push(fullPath);
-                continue;
+
+                let rootReason = ""
```
Comment (dcrousso) [inline Source/WebInspectorUI/UserInterface/Workers/HeapSnapshot/HeapSnapshot.js]: Style: missing `;`

ditto below

File: Source/WebInspectorUI/UserInterface/Workers/HeapSnapshot/HeapSnapshot.js [hunk 12]
```diff
@@ -761,11 +773,26 @@ HeapSnapshot = class HeapSnapshot
...
+
+                for (let rootIndex = 0; rootIndex < this._roots.length; rootIndex += rootFieldCount) {
+                    let nodeID = snapshot._roots[rootIndex + rootNodeIdOffset]
+                    let reason = snapshot._labels[snapshot._roots[rootIndex + rootReasonIndexOffset]]
```
Comment (dcrousso) [inline Source/WebInspectorUI/UserInterface/Workers/HeapSnapshot/HeapSnapshot.js]: where is `snapshot` defined?

File: Source/WebInspectorUI/UserInterface/Workers/HeapSnapshot/HeapSnapshot.js [hunk 13]
```diff
@@ -761,11 +773,26 @@ HeapSnapshot = class HeapSnapshot
...
+                    if (this._nodeIdentifierToOrdinal.get(nodeID) != nodeOrdinal)
```
Comment (dcrousso) [inline Source/WebInspectorUI/UserInterface/Workers/HeapSnapshot/HeapSnapshot.js]: Style: `!==`

ditto below

File: Source/JavaScriptCore/tools/JSDollarVM.cpp (hunk 3 above)
Comment (darinadler) [reply Source/JavaScriptCore/tools/JSDollarVM.cpp]: It is alphabetical already. The "w" from "wtf" comes after "u" from "unicode".

File: Source/cmake/OptionsGTK.cmake (hunk 2 above)
Comment (kmiller68) [reply Source/cmake/OptionsGTK.cmake]: > > The Apple ports enable this for debug builds automatically now
> 
> No they don't. A `git grep` indicates `ENABLE_JSC_RESTRICTED_OPTIONS_BY_DEFAULT` does not exist. :)

Yeah, it's there, in `Source/JavaScriptCore/Configurations/DebugRelease.xcconfig`:
```
GCC_PREPROCESSOR_DEFINITIONS = $(inherited) ENABLE_JSC_RESTRICTED_OPTIONS_BY_DEFAULT=1;
```

File: Source/JavaScriptCore/heap/Strong.h (hunk 6 above)
Comment (justinmichaud) [reply Source/JavaScriptCore/heap/Strong.h]: I did investigate this, the issue was that templated classes can't really export static members to other modules. Fixed!

File: Source/JavaScriptCore/tools/JSDollarVM.cpp [hunk 14]
```diff
@@ -2533,6 +2536,26 @@ JSC_DEFINE_HOST_FUNCTION(functionGC, (JSGlobalObject* globalObject, CallFrame*))
...
 }
 
+// Runs a full GC synchronously.
+// Usage: $vm.triggerMemoryPressure()
+JSC_DEFINE_HOST_FUNCTION(functionTriggerMemoryPressure, (JSGlobalObject* globalObject, CallFrame*))
+{
+    DollarVMAssertScope assertScope;
+    dataLogLn("functionTriggerMemoryPressure: ", Options::dumpHeapOnLowMemory(), " ", Options::enableStrongRefTracker());
+    globalObject->vm().drainMicrotasks();
+    WTF::MemoryPressureHandler::singleton().endSimulatedMemoryPressure();
```
Comment (rreno) [inline Source/JavaScriptCore/tools/JSDollarVM.cpp]: These begin/end pairings look reversed to me.

File: Source/WTF/wtf/MemoryPressureHandler.cpp [hunk 15]
```diff
@@ -85,12 +85,6 @@ void MemoryPressureHandler::setMemoryFootprintPollIntervalForTesting(Seconds pol
 
 void MemoryPressureHandler::setShouldUsePeriodicMemoryMonitor(bool use)
 {
-    if (!isFastMallocEnabled()) {
```
Comment (rreno) [inline Source/WTF/wtf/MemoryPressureHandler.cpp:88]: What's the reason for this change? I think this will break malloc stack logging on Apple platforms.

File: Source/WTF/wtf/RefTrackerMixin.h [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+    WTF_EXPORT_PRIVATE void logAllLiveReferences();
+
+    Lock lock { };
+    HashMap<void*, std::unique_ptr<StackShot>> map { };
+    std::atomic<int> loggingDisabledDepth { };
+};
+
+template<typename T>
+struct RefTrackerMixin final {
+    ALWAYS_INLINE explicit RefTrackerMixin()
```
Comment (rreno) [inline Source/WTF/wtf/RefTrackerMixin.h]: Is `explicit` required for zero-argument constructors?

File: Source/WTF/wtf/RefTrackerMixin.h [hunk 17]
```diff
@@ -0,0 +1,152 @@
...
+public:
+    RefTracker() = default;
+    ~RefTracker() = default;
+
+    // NEVER_INLINE to make skipping frames more predictable.
+    WTF_EXPORT_PRIVATE void reportLive(void*);
+    WTF_EXPORT_PRIVATE void reportDead(void*);
+    WTF_EXPORT_PRIVATE void logAllLiveReferences();
+
+    Lock lock { };
```
Comment (rreno) [inline Source/WTF/wtf/RefTrackerMixin.h:55]: Could you annotate which data structure is guarded by the lock with `WTF_GUARDED_BY_LOCK` for the static analyzer, please?

File: Source/WTF/wtf/RefTrackerMixin.h [hunk 18]
```diff
@@ -0,0 +1,152 @@
...
+        WTF_EXPORT_PRIVATE static WTF::RefTracker& refTrackerSingleton(); \
+    };
+
+#define REFTRACKER_MEMBERS(T) \
+    WTF::RefTrackerMixin<T> m_refTrackerData;
+
+#define REFTRACKER_IMPL(T) \
+    WTF::RefTracker& T::refTrackerSingleton() \
+    { \
+        static WTF::RefTracker* s_singleton; \
```
Comment (rreno) [inline Source/WTF/wtf/RefTrackerMixin.h]: I think this could be a `static NeverDestroyed<RefTracker>`.

File: Source/WTF/wtf/RefTrackerMixin.h (hunk 17 above)
Comment (rreno) [reply Source/WTF/wtf/RefTrackerMixin.h:55]: I mean yeah it's obvious which one it is but it would be nice to get the compilation failures for the lack of holding the lock when using the map.

File: Source/JavaScriptCore/runtime/VM.cpp [hunk 19]
```diff
@@ -372,7 +372,12 @@ VM::VM(VMType vmType, HeapType heapType, WTF::RunLoop* runLoop, bool* success)
     // won't use this.
     m_typedArrayController = adoptRef(new SimpleTypedArrayController());
 
-    m_bytecodeIntrinsicRegistry = makeUnique<BytecodeIntrinsicRegistry>(*this);
+    {
+#if ENABLE(REFTRACKER)
+        WTF::RefTrackerLoggingDisabledScope<StrongRefTracker> disableLoggingScope;
```
Comment (rreno) [inline Source/JavaScriptCore/runtime/VM.cpp]: Why do we do this here?

File: Source/WTF/wtf/MemoryPressureHandler.cpp (hunk 15 above)
Comment (rreno) [reply Source/WTF/wtf/MemoryPressureHandler.cpp:88]: Well, it won't break it but it might make that mode more prone to memory kills.

Comment (justinmichaud) [reply Source/WTF/wtf/MemoryPressureHandler.cpp:88]: I moved this check to the caller (well, I guess only for the web content process) so that it can check JSC options too.

I can add it in for the other processes too if they might get frequently killed

File: Source/WebKit/WebProcess/WebProcess.cpp [hunk 20]
```diff
@@ -470,7 +473,11 @@ void WebProcess::initializeWebProcess(WebProcessCreationParameters&& parameters,
 #if ENABLE(PERIODIC_MEMORY_MONITOR)
         if (auto pollInterval = parameters.memoryFootprintPollIntervalForTesting)
             memoryPressureHandler.setMemoryFootprintPollIntervalForTesting(pollInterval);
-        memoryPressureHandler.setShouldUsePeriodicMemoryMonitor(true);
+#if !USE(SYSTEM_MALLOC)
+        // If we're running with FastMalloc disabled, some kind of testing or debugging is probably happening.
```
Comment (justinmichaud) [inline Source/WebKit/WebProcess/WebProcess.cpp:477]: @rreno This check was moved to here

File: Source/JavaScriptCore/runtime/VM.cpp (hunk 19 above)
Comment (justinmichaud) [reply Source/JavaScriptCore/runtime/VM.cpp]: I found that this cache was generating a bunch of noise. But it isn't needed

File: Source/WTF/wtf/MemoryPressureHandler.cpp (hunk 15 above)
Comment (rreno) [reply Source/WTF/wtf/MemoryPressureHandler.cpp:88]: Ah I missed that, thanks.

File: Source/JavaScriptCore/heap/Strong.h [hunk 21]
```diff
@@ -70,15 +73,24 @@ template <typename T, ShouldStrongDestructorGrabLock shouldStrongDestructorGrabL
...
-    ~Strong()
+    enum HashTableEmptyValueTag { HashTableEmptyValue };
+    bool isHashTableEmptyValue() const { return slot() == hashTableEmptyValue(); }
+
+    Strong(HashTableEmptyValueTag)
+        : Handle<T>(hashTableEmptyValue())
+    {
+    }
+
+    virtual ~Strong()
```
Comment (ddegazio) [inline Source/JavaScriptCore/heap/Strong.h]: I think this should probably not be `virtual`, and instead `~Strong() override`.

File: Source/JavaScriptCore/heap/Strong.h [hunk 22]
```diff
@@ -157,8 +172,31 @@ namespace WTF {
...
+    static constexpr bool canInitializeWithMemset = false;
+    static constexpr bool canMoveWithMemcpy = false;
+#endif
 };
 
-template<typename P> struct HashTraits<JSC::Strong<P>> : SimpleClassHashTraits<JSC::Strong<P>> { };
+template<typename P> struct HashTraits<JSC::Strong<P>> : SimpleClassHashTraits<JSC::Strong<P>> {
+#if ENABLE(REFTRACKER)
+    using S = JSC::Strong<P>;
+    static constexpr bool emptyValueIsZero = false;
```
Comment (ddegazio) [inline Source/JavaScriptCore/heap/Strong.h:184]: Why isn't `emptyValueIsZero` true? `Strong::hashTableEmptyValue()` as defined above returns a null pointer, and I don't think `Strong` nor `Handle` define other fields.

**Review by ddegazio — APPROVED** [review]
r=me

File: Source/JavaScriptCore/heap/Strong.h (hunk 22 above)
Comment (justinmichaud) [reply Source/JavaScriptCore/heap/Strong.h:184]: if constexpr (Traits::emptyValueIsZero)
        return reinterpret_cast_ptr<ValueType*>(static_cast<char*>(HashTableMalloc::zeroedMalloc(size * sizeof(ValueType))));
        
  If we allocate a bucket but don't call Strong(), then set it later, then the ref tracker will never find out about it. To be honest, it can probably be made to work without this, but I was having a very difficult time figuring out why I was seeing a ref that was created but never freed, so I just tried to be conservative.
